specs/SharedBuffer.md

```diff
@@ -0,0 +1,282 @@
+      }
+      else
+      {
+           m_webview.ShareBufferToScript(sharedBuffer, /*isReadOnlyToScript*/true, additionalDataAsJson);
+      }
+      // Explicitly close the one time shared buffer to ensure that the resource is released timely.
+      sharedBuffer.Close();
+```
+
+# API Details
+## Win32 C++
+```
+interface ICoreWebView2StagingEnvironment2 : IUnknown {
+  /// Create a shared memory based buffer with the specifid size in bytes.
+  /// The buffer can be shared with web contents in WebView by calling
+  /// `ShareBufferToScript` on `CoreWebView2` or `CoreWebViewFrame` object.
+  /// Once shared, the same content of the buffer will be accessible from both
+  /// the app process and script in WebView. Modification to the content will be visible
+  /// to all parties that have access to the buffer.
+  /// For 32bit application, the creation will fail with E_INVALIDARG if `size` is larger than 4GB.
+  HRESULT CreateSharedBuffer(
+    [in] UINT64 size,
+    [out, retval] ICoreWebView2StagingSharedBuffer** shared_buffer);
+}
+
+interface ICoreWebView2StagingSharedBuffer : IUnknown {
+  /// The size of the shared buffer in bytes.
+  [propget] HRESULT Size([out, retval] UINT64* value);
+
+  /// The memory address of the shared buffer.
+  [propget] HRESULT Buffer([out, retval] BYTE** value);
+
+  /// Get an IStream object that can be used to access the shared buffer.
+  HRESULT GetAccessStream([out, retval] IStream** value);
+
+  /// The file mapping handle of the shared memory of the buffer.
+  /// Normal app should use `Buffer` or `GetAccessStream` to get memory address
+  /// or IStream object to access the buffer.
+  /// For advanced scenario, you could duplicate this handle to another application
+  /// process and create a mapping from the duplicated handle in that process to access
+  /// the buffer from that separate process.
+  [propget] HRESULT Handle([out, retval] HANDLE* value);
+  
+  /// Release the backing shared memory. The application should call this API when no
+  /// access to the buffer is needed any more, to ensure that the underlying resources
+  /// are released timely even if the shared buffer object itself is not released due to some leaked
+  /// reference.
+  /// After the shared buffer is closed, accessing properties of the object will fail with
+  /// `HRESULT_FROM_WIN32(ERROR_INVALID_STATE)`. Operations like Read or Write on the IStream objects
+  /// returned from `GetAccessStream` will fail with `HRESULT_FROM_WIN32(ERROR_INVALID_STATE)`.
+  /// `ShareBufferToScript` will also fail with `HRESULT_FROM_WIN32(ERROR_INVALID_STATE)`.
+  ///
+  /// The script code should call `chrome.webview.releaseSharedBuffer` with
+  /// the shared buffer as the parameter to release underlying resources as soon
+  /// as it does not need access the shared buffer any more.
+  /// When script tries to access the buffer after calling `chrome.webview.releaseSharedBuffer`,
```

Comment:
The Close on native side only impacts application process's access to the buffer while the releaseSharedBuffer only affects script's access to the buffer. The underlying OS shared memory will only be release when both sides release the buffer.
Will update the spec to make it clear.

Reply:
I thought this would act like IDisposable/IClosable where you force the close of the underlying data regardless of if anyone has a reference to it. We already have the cleanup once everything is done using it via just not calling Close right and relying on destructors right? If we add an explicit Close method shouldn't it immediately cleanup the shared memory?

Reply:
This is something different because the object is shared. It's like you can Close a HANDLE to a file as part of some disposal, but the underlying file will only be really closed by the OS when all HANDLEs including the duplicated ones are closed.
It is also much easier to work with if native side and script side can independently close/release the buffer. Otherwise, it would require coordination, which is hard.

Reply:
OK saying its closing its HANDLE makes sense to me now =)